### PR TITLE
added option to write all repl prints to a logfile

### PR DESCRIPTION
**Pull request by r-e-l-z**
I find it useful to record the Frida output to a logfile, especially for long runs. The repl _resists_ regular means such as tee and redirection, so I added this option.

### CODE REVIEW

File: src/frida/application.py [hunk 1]
```diff
@@ -119,6 +121,15 @@ def store_target(option, opt_str, target_value, parser, target_type, *args, **kw
         self._exit_status = None
         self._console_state = ConsoleState.EMPTY
         self._quiet = False
+        self._logfile = options.logfile
+
+        if self._logfile:
```
Comment (oleavr) [inline src/frida/application.py]: For consistency:
`if self._logfile is not None:`

File: src/frida/application.py [hunk 2]
```diff
@@ -119,6 +121,15 @@ def store_target(option, opt_str, target_value, parser, target_type, *args, **kw
...
+            try:
+                self._f = open(self._logfile, 'w')
+            except Exception as e:
+                self._update_status('Failed to open logfile "%s"' % self._logfile)
+                sys.exit(1)
+
```
Comment (oleavr) [inline src/frida/application.py]: Leftover newline.

File: src/frida/application.py [hunk 3]
```diff
@@ -265,6 +276,7 @@ def _on_output(self, pid, fd, data):
         lines = text.split("\n")
         self._print(prefix + ("\n" + prefix).join(lines))
 
+
```
Comment (oleavr) [inline src/frida/application.py]: Leftover newline.

File: src/frida/application.py [hunk 4]
```diff
@@ -302,6 +314,7 @@ def _print(self, *args, **kwargs):
             else:
                 encoded_args.append(arg)
         print(*encoded_args, **kwargs)
+
```
Comment (oleavr) [inline src/frida/application.py]: Leftover newline.

File: src/frida/application.py [hunk 5]
```diff
@@ -310,6 +323,8 @@ def _log(self, level, text):
         else:
             color = Fore.RED if level == 'error' else Fore.YELLOW
             self._print(color + Style.BRIGHT + text + Style.RESET_ALL)
+        if self._logfile:
+            self._f.write(text + "\n");
```
Comment (oleavr) [inline src/frida/application.py]: For consistency:
`if self._logfile is not None:`

File: src/frida/application.py [hunk 6]
```diff
@@ -119,7 +120,13 @@ def store_target(option, opt_str, target_value, parser, target_type, *args, **kw
         self._exit_status = None
         self._console_state = ConsoleState.EMPTY
         self._quiet = False
-
+        self._logfile = options.logfile
+        if self._logfile is not None:
+            try:
+                self._f = open(self._logfile, 'w')
```
Comment (oleavr) [inline src/frida/application.py]: I missed one thing, sorry, could we use `options.logfile` directly here in the `if` and `open`, and replace `_f` by `_logfile`? The logfile path isn't used again, so we don't have to store it on the object, and `_logfile` leaves us with a more descriptive name than `_f`.

**Comment by r-e-l-z** [comment]
Yeah, that makes sense. I'll fix it and re-push.

**Comment by oleavr** [comment]
Thanks! I think we also need to set it to `None` in the case where there's no logfile specified -- did you test if this still works?

**Comment by r-e-l-z** [comment]
Sorry, about that. Forgot to test it without a logfile, since I'm using a log file all the time now...

**Comment by r-e-l-z** [comment]
I just found it out myself, before reading your comment, when I finally did a run without a log file :)

**Comment by oleavr** [comment]
No worries! :) Great stuff, let's get this landed.

**Comment by oleavr** [comment]
Oops, I didn't notice until after merging that we added this logic to the ConsoleApplication base class shared by frida, frida-trace, etc., so I just moved this over to the REPL:
https://github.com/frida/frida-python/commit/89340c1526806c5a9b31d256779284db1971ae92

**Pull request merged**